### PR TITLE
Turn widget into a provider

### PR DESCRIPTION
**Pull request by seromenho**
- Stores directly on app settings
- Added new custom condition
- Added new Data Source as a requirement
- Possibility to add/edit multiple hooks for the app

### CODE REVIEW

File: js/interface.js [hunk 1]
```diff
@@ -11,115 +11,247 @@ var rule = {
...
-  $('#name').val(hooksNames[0]);
-  $('#filterType').val(rule.filterType);
-  $('#requirement').val(rule.requirement);
-  $('#errorMessage').val(rule.errorMessage);
-}
+updateSelectText($('[data-name="requirement"]'));
+updateSelectText($('[data-name="filterType"]'));
+updateSelectText($('[data-name="hookType"]'));
+
+var i = 1; // Just a counter to have a simple unique id for each hook when rendering the template.
```
Comment (squallstar) [inline js/interface.js]: then why not calling this `templateUniqueId` 😂  or similar

File: js/interface.js [hunk 2]
```diff
@@ -11,115 +11,247 @@ var rule = {
...
 });
 
-onErrorActionProvider.then(function (result) {
-  rule.onErrorAction = result && result.data;
-  $('form').submit();
-})
+function addHookItem(settings) {
+  settings = settings || {};
+  settings.pages = settings.pages || pages;
+  i = i + 1;
```
Comment (squallstar) [inline js/interface.js]: `i++`

File: js/interface.js [hunk 3]
```diff
@@ -11,115 +11,247 @@ var rule = {
...
-      .then(function () {
-        Fliplet.Widget.complete();
-      })
-      .catch(function(reason) {
-        console.log(reason);
-      });
-    });
-  });
+$('[data-name="filterType"]').on('change', function () {
+
```
Comment (squallstar) [inline js/interface.js]: looks empty?

File: js/interface.js [hunk 4]
```diff
@@ -11,115 +11,247 @@ var rule = {
...
+    newHooks[hookName].run = [$(this).find('#filterType').val()];
+    
+    onErrorActionProviders[id].then(function (result) {
+      newHooks[hookName].settings.onErrorAction = result && result.data;
+      newHooks[hookName].script = compile(newHooks[hookName].settings);
+    });
+  });
+
+  Promise.all(Object.keys(onErrorActionProviders).map(function(id) {
+    return onErrorActionProviders[id].forwardSaveRequest();
```
Comment (squallstar) [inline js/interface.js]: I don't think this returns a promise

File: js/interface.js (hunk 1 above)
Comment (seromenho) [inline js/interface.js]: So long that name 👍 
But obviously I agree 😂

File: js/interface.js (hunk 2 above)
Comment (seromenho) [inline js/interface.js]: no-plusplus
i += 1; 😄

File: js/interface.js (hunk 4 above)
Comment (seromenho) [inline js/interface.js]: Ah. True. Changing it.

File: js/interface.js (hunk 3 above)
Comment (seromenho) [inline js/interface.js]: To be removed

**Comment by hcarneiro** [comment]
@seromenho Here is a list of issues I have found that is best for you to look into it:
- If you have a list of screen already saved in one rule, creating a new item, will come with those same screens selected by default
- Link widget disappears on save
- Link widget settings don’t load after being saved

There's also the reorder thing for you. The UI is now capable of drag & drop.

**Comment by hcarneiro** [comment]
@seromenho With regards to the issue of *Link widget settings don’t load after being saved*, it seems that this:
```
return Fliplet.API.request({
  url: 'v1/apps/' + Fliplet.Env.get('appId'),
  method: 'PUT',
  data: {
    hooks: newHooks
  }
});
```
Is being run before we save the link data into the hooks objects.

**Pull request merged**